### ISSUE TITLE
Deallocate correctly to avoid memory leaks

### ISSUE THREAD

**Issue by czgdp1807** [lfortran/lfortran#2635]
This PR deallocates before allocating. However since that needs an extra check so @certik suggested to do it at the end of iteration if allocate happens inside a loop. That's tricky so I am delaying that for another PR.

_Originally posted by @czgdp1807 in https://github.com/lfortran/lfortran/issues/2615#issuecomment-1764076235_

**Comment by czgdp1807** [comment]
@certik suggested whatever we have is good. So this issue can wait and is low priority.

**Comment by certik** [comment]
Yes, this issue is low priority --- we can even consider fixing this via an ASR pass(es) that will optimize allocatables.

A quick optimization might be to check if we are not in a loop and in that case not to insert the ExplicitDeallocate.
